File: competitive-programming/src/edu/algos/li04_hashing/medium/L128_LongestConsecutiveSequence.java
package edu.algos.li04_hashing.medium;

import java.util.HashSet;

public class L128_LongestConsecutiveSequence {

	public static void main(String[] args) {
		L128_LongestConsecutiveSequence obj = new L128_LongestConsecutiveSequence();
		System.out.println(obj.longestConsecutive(new int[] { 100, 4, 200, 1, 3, 2 }));
	}

	// Problem: https://leetcode.com/problems/longest-consecutive-sequence/
	// Idea: Use HashSet to check if num-1 exist in the set.
	public int longestConsecutive(int[] nums) {
		HashSet<Integer> set = new HashSet<Integer>();
		for (int i = 0; i < nums.length; i++) {
			set.add(nums[i]);
		}

		int max = 0;
		for (int i = 0; i < nums.length; i++) {
			int current = nums[i];
			if (set.contains(current - 1))
				continue;

			int count = 0;
			while (set.contains(current)) {
				count++;
				current++;
			}
			if (count > max) {
				max = count;
			}
		}
		return max;
	}
}

File: dsa/topics/02_hashing/python/l128_longest_consecutive_sequence.py
# L128 - Longest Consecutive Sequence
#
# Problem: https://leetcode.com/problems/longest-consecutive-sequence/
#
# Idea:
# - Store all numbers in a set
# - Only start counting when (num - 1) is not present
#
# Time: O(n)
# Space: O(n)


def longest_consecutive(nums):
    if not nums:
        return 0

    num_set = set(nums)
    longest = 0

    for num in num_set:  # Numbers can be repeating in nums. Better to traverse unique nums.
        # Start only if this number is the beginning of a sequence
        if num - 1 not in num_set:
            current = num
            length = 1

            while current + 1 in num_set:
                current += 1
                length += 1

            longest = max(longest, length)

    return longest


# -------------------- TEST --------------------
if __name__ == "__main__":
    nums = [100, 4, 200, 1, 3, 2]
    print(longest_consecutive(nums))  # Output: 4
